gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BomConflictCheckTask.java

```diff
@@ -0,0 +1,136 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+import netflix.nebula.dependency.recommender.provider.RecommendationProviderContainer;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.TaskAction;
+
+
+public class BomConflictCheckTask extends DefaultTask {
+
+    @TaskAction
+    public final void checkBomConflict() {
+        final Map<String, String> recommendations = getProject().getExtensions()
+                .getByType(RecommendationProviderContainer.class)
+                .getMavenBomProvider()
+                .getRecommendations();
+        List<Conflict> conflicts = new LinkedList<>();
+        Set<String> artifacts = BaselineVersions.getResolvedArtifacts(getProject());
+        Map<String, String> resolvedConflicts = new HashMap<>();
+        BaselineVersions.checkVersionsProp(getProject(),
+                pair -> {
+                    String propName = pair.getLeft();
+                    String propVersion = pair.getRight();
+                    String regex = propName.replaceAll("\\*", ".*");
+                    artifacts.forEach(artifactName -> {
+                        if (artifactName.matches(regex)) {
+                            resolvedConflicts.put(artifactName, propName);
+                        }
+                    });
+                    recommendations.forEach((bomName, bomVersion) -> {
+                        if (bomName.matches(regex)) {
+                            conflicts.add(new Conflict(propName, propVersion, bomName, bomVersion));
+                            resolvedConflicts.remove(bomName);
+                        }
+                    });
+                    return null;
+                });
+
+        //Critical conflicts are versions.props line that only override bom recommendations with same version
+        //so it should avoid considering the case where a wildcard also pin an artifact not present in the
+        List<Conflict> critical = conflicts.stream()
+                .filter(c -> c.getBomVersion().equals(c.getPropVersion()))
+                .filter(c -> !resolvedConflicts.containsValue(c.getPropName()))
+                .collect(Collectors.toList());
+
+        if (!conflicts.isEmpty()) {
+            System.out.println("There are conflicts between versions.props and the bom:");
+            System.out.println(conflictsToString(conflicts, resolvedConflicts));
+
+            if (!critical.isEmpty()) {
+                throw new RuntimeException("Critical conflicts between versions.props and the bom "
+                        + "(overriding with same version)\n" + conflictsToString(critical, resolvedConflicts));
+            }
+        }
+
+    }
+
+    private String conflictsToString(List<Conflict> conflicts, Map<String, String> resolvedConflicts) {
+        return conflicts.stream()
+                .map(conflict -> conflict.details(resolvedConflicts))
+                .collect(Collectors.joining("\n"));
+    }
+
+    private static class Conflict {
```

Comment:
use immutables for this?

Reply:
It's not already a dependency, not sure I want to add it as a dependency for removing a few lines of boilerplate

Reply:
Fair